File: OfflineFetcher/app/src/main/java/de/jbamberger/offlinefetcher/source/reddit/RedditRepository.java
package de.jbamberger.offlinefetcher.source.reddit;

import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;

import java.util.concurrent.Executor;

import javax.inject.Singleton;

import de.jbamberger.offlinefetcher.db.dao.RedditPostDao;
import de.jbamberger.offlinefetcher.db.entity.RedditPost;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * @author Jannik Bamberger (dev.jbamberger@gmail.com)
 */
@Singleton
public class RedditRepository {
    private final RedditApi api;
    private final Executor executor;
    private final RedditPostDao redditPostDao;

    public RedditRepository(RedditApi api, Executor executor, RedditPostDao redditPostDao) {
        this.api = api;
        this.executor = executor;
        this.redditPostDao = redditPostDao;
    }

    public LiveData<RedditPost> getRedditPost(String subreddit) {
        refreshRedditPosts(subreddit);
        return redditPostDao.load(subreddit);
    }

    private void refreshRedditPosts(final String subreddit) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                boolean postExists = false;//redditPostDao.hasRedditPost(TIMEOUT);
                if (!postExists) {
                    api.getPosts(subreddit).enqueue(new Callback<RedditPost>() {
                        @Override
                        public void onResponse(@NonNull Call<RedditPost> call, @NonNull Response<RedditPost> response) {
                            // error case is left out for brevity
                            redditPostDao.save(response.body());
                        }

                        @Override
                        public void onFailure(@NonNull Call<RedditPost> call, @NonNull Throwable t) {

                        }
                    });
                }
            }
        });
    }
}

File: OfflineClient/cached-api/src/main/java/de/jbamberger/api/provider/reddit/RedditRepository.kt
package de.jbamberger.api.provider.reddit

import android.arch.lifecycle.LiveData
import de.jbamberger.api.AppExecutors
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import javax.inject.Inject
import javax.inject.Singleton

/**
 * @author Jannik Bamberger (dev.jbamberger@gmail.com)
 */
@Singleton
class RedditRepository @Inject internal constructor(
        private val api: RedditApi,
        private val executor: AppExecutors,
        private val redditPostDao: RedditPostDao) {

    fun getRedditPost(subreddit: String): LiveData<RedditPost> {
        refreshRedditPosts(subreddit)
        return redditPostDao.load(subreddit)
    }

    private fun refreshRedditPosts(subreddit: String) {
        executor.networkIO().execute {
            val postExists = false//redditPostDao.hasRedditPost(TIMEOUT);
            if (!postExists) {
                api.getPosts(subreddit).enqueue(object : Callback<RedditPost> {
                    override fun onResponse(call: Call<RedditPost>, response: Response<RedditPost>) {
                        // error case is left out for brevity
                        executor.diskIO().execute { redditPostDao.save(response.body()!!) }
                    }

                    override fun onFailure(call: Call<RedditPost>, t: Throwable) {

                    }
                })
            }
        }
    }
}
